config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java

```diff
@@ -285,7 +285,10 @@ private static boolean hasAtLeastOneBeanProperty(Class<?> clazz) {
 
     private static boolean isOptionalProperty(Class beanClass, PropertyDescriptor beanProp) {
         Field field = getField(beanClass, beanProp.getName());
-        return field != null ? field.getAnnotationsByType(Optional.class).length > 0 : beanProp.getReadMethod().getAnnotationsByType(Optional.class).length > 0;
+        return field != null
+                ? field.getAnnotationsByType(Optional.class).length > 0
+                        || beanProp.getReadMethod().getAnnotationsByType(Optional.class).length > 0
```

Comment:
Maybe 
`beanProp.getReadMethod().getAnnotationsByType(Optional.class).length > 0` should be checked before calling getField ?
Method getField modifies field via reflection and optionally recursively calls itself, so it could be costly.